File: JavaScript/Interview_problems/Chess_Game.js
/*
================================================================================
LLD: Chess Game
================================================================================

Functional Requirements:
    1. Support a standard 8x8 board.
    2. Enforce piece-specific movement and capture rules for every piece type
       (KING, QUEEN, ROOK, BISHOP, KNIGHT, PAWN).
    3. Detect check, checkmate, and stalemate after every move.
    4. Maintain move history to support undo.

Non-Functional Requirements:
    1. Move validation latency should be small: each call generates a
       piece's candidate moves in O(board size), never scanning the whole
       game tree.
    2. Extensibility: adding a new piece type or a variant rule (e.g.
       castling) should not require touching unrelated piece classes.

--------------------------------------------------------------------------
Design
--------------------------------------------------------------------------
Strategy pattern for movement:
    Every piece knows only its OWN movement rule (Piece#getPossibleMoves).
    Board/Game never special-case "if this is a bishop do X" -- they just
    call piece.getPossibleMoves(...) polymorphically. Adding a new piece
    type means adding one new subclass, nothing else changes.

Command-style Move objects for undo:
    Each executed Move records exactly what it takes to reverse itself: the
    piece that was AT the source square before the move (so promotion can
    be undone back to a pawn) and whatever piece occupied the destination
    square before it was captured (possibly null). undo() pops the last
    Move and replays those two facts backwards -- it never needs to
    recompute anything from scratch.

Core Entities:
    Game             -- orchestrates turns, applies/undoes moves, tracks status
    Board            -- 8x8 grid of Piece references + board-level queries
    Piece (abstract)  -- color, type, getPossibleMoves() [King, Queen,
                         Rook, Bishop, Knight, Pawn]
    Move             -- source, destination, pieceMoved (pre-move), capturedPiece
    Player           -- name + color
    GameStatus       -- ONGOING, CHECK, CHECKMATE, STALEMATE

Out of scope: castling, en passant, draw by repetition/50-move rule.
================================================================================
*/

'use strict';

const Color = Object.freeze({ WHITE: 'WHITE', BLACK: 'BLACK' });

function opposite(color) {
    return color === Color.WHITE ? Color.BLACK : Color.WHITE;
}

const PieceType = Object.freeze({
    KING: 'KING', QUEEN: 'QUEEN', ROOK: 'ROOK', BISHOP: 'BISHOP', KNIGHT: 'KNIGHT', PAWN: 'PAWN',
});

const GameStatus = Object.freeze({
    ONGOING: 'ONGOING', CHECK: 'CHECK', CHECKMATE: 'CHECKMATE', STALEMATE: 'STALEMATE',
});

// A Position is just a {row, col} pair; row 0 = rank 1 (White's back rank),
// col 0 = file 'a'. Two positions are compared by value with samePosition
// since plain object identity would never match across calls.
function samePosition(a, b) {
    return a.row === b.row && a.col === b.col;
}

class Piece {
    constructor(color, type) {
        this.color = color;
        this.type = type;
    }

    // Pseudo-legal moves: obeys this piece's movement/blocking/capture
    // rules, but ignores whether the move leaves the mover's own king in
    // check -- Game.move() layers that safety check on top.
    // eslint-disable-next-line no-unused-vars
    getPossibleMoves(position, board) {
        throw new Error('getPossibleMoves must be implemented by subclass');
    }

    // Squares this piece threatens; used for check detection. Identical to
    // getPossibleMoves for every piece except the pawn, whose diagonal
    // attacks threaten a square whether or not an enemy is standing on it
    // yet (a plain move-list wouldn't include an empty diagonal square).
    getAttackedSquares(position, board) {
        return this.getPossibleMoves(position, board);
    }

    symbol() {
        const letters = {
            KING: 'K', QUEEN: 'Q', ROOK: 'R', BISHOP: 'B', KNIGHT: 'N', PAWN: 'P',
        };
        const letter = letters[this.type];
        return this.color === Color.WHITE ? letter : letter.toLowerCase();
    }

    // Shared helper for the sliding pieces (rook/bishop/queen): walk each
    // direction until the edge of the board, a friendly piece (stop before
    // it), or an enemy piece (stop after capturing it).
    _slide(position, board, directions) {
        const moves = [];
        for (const [dRow, dCol] of directions) {
            let r = position.row + dRow;
            let c = position.col + dCol;
            while (board.isValidCell(r, c)) {
                const occupant = board.getPiece(r, c);
                if (!occupant) {
                    moves.push({ row: r, col: c });
                } else {
                    if (occupant.color !== this.color) {
                        moves.push({ row: r, col: c });
                    }
                    break;
                }
                r += dRow;
                c += dCol;
            }
        }
        return moves;
    }
}

class Rook extends Piece {
    constructor(color) {
        super(color, PieceType.ROOK);
    }

    getPossibleMoves(position, board) {
        return this._slide(position, board, [[1, 0], [-1, 0], [0, 1], [0, -1]]);
    }
}

class Bishop extends Piece {
    constructor(color) {
        super(color, PieceType.BISHOP);
    }

    getPossibleMoves(position, board) {
        return this._slide(position, board, [[1, 1], [1, -1], [-1, 1], [-1, -1]]);
    }
}

class Queen extends Piece {
    constructor(color) {
        super(color, PieceType.QUEEN);
    }

    getPossibleMoves(position, board) {
        return this._slide(position, board, [
            [1, 0], [-1, 0], [0, 1], [0, -1], [1, 1], [1, -1], [-1, 1], [-1, -1],
        ]);
    }
}

const KNIGHT_OFFSETS = [[2, 1], [2, -1], [-2, 1], [-2, -1], [1, 2], [1, -2], [-1, 2], [-1, -2]];

class Knight extends Piece {
    constructor(color) {
        super(color, PieceType.KNIGHT);
    }

    getPossibleMoves(position, board) {
        const moves = [];
        for (const [dRow, dCol] of KNIGHT_OFFSETS) {
            const r = position.row + dRow;
            const c = position.col + dCol;
            if (board.isValidCell(r, c)) {
                const occupant = board.getPiece(r, c);
                if (!occupant || occupant.color !== this.color) {
                    moves.push({ row: r, col: c });
                }
            }
        }
        return moves;
    }
}

class King extends Piece {
    constructor(color) {
        super(color, PieceType.KING);
    }

    getPossibleMoves(position, board) {
        const moves = [];
        for (let dRow = -1; dRow <= 1; dRow++) {
            for (let dCol = -1; dCol <= 1; dCol++) {
                if (dRow === 0 && dCol === 0) continue;
                const r = position.row + dRow;
                const c = position.col + dCol;
                if (board.isValidCell(r, c)) {
                    const occupant = board.getPiece(r, c);
                    if (!occupant || occupant.color !== this.color) {
                        moves.push({ row: r, col: c });
                    }
                }
            }
        }
        return moves;
    }
}

class Pawn extends Piece {
    constructor(color) {
        super(color, PieceType.PAWN);
    }

    getPossibleMoves(position, board) {
        const direction = this.color === Color.WHITE ? 1 : -1;
        const startRow = this.color === Color.WHITE ? 1 : 6;
        const moves = [];

        const oneStep = { row: position.row + direction, col: position.col };
        if (board.isValidCell(oneStep.row, oneStep.col) && !board.getPiece(oneStep.row, oneStep.col)) {
            moves.push(oneStep);
            const twoStep = { row: position.row + 2 * direction, col: position.col };
            if (position.row === startRow && !board.getPiece(twoStep.row, twoStep.col)) {
                moves.push(twoStep);
            }
        }

        for (const dCol of [-1, 1]) {
            const r = position.row + direction;
            const c = position.col + dCol;
            if (board.isValidCell(r, c)) {
                const occupant = board.getPiece(r, c);
                if (occupant && occupant.color !== this.color) {
                    moves.push({ row: r, col: c });
                }
            }
        }
        return moves;
    }

    getAttackedSquares(position, board) {
        const direction = this.color === Color.WHITE ? 1 : -1;
        const squares = [];
        for (const dCol of [-1, 1]) {
            const r = position.row + direction;
            const c = position.col + dCol;
            if (board.isValidCell(r, c)) {
                squares.push({ row: r, col: c });
            }
        }
        return squares;
    }
}

// A Move is a self-contained "command": it records the two facts needed to
// reverse itself (pieceMoved -- as it was BEFORE this move, so a promoted
// queen can be undone back into a pawn -- and capturedPiece, possibly
// null), so Game#undo never has to recompute board state from scratch.
class Move {
    constructor(source, destination, pieceMoved, capturedPiece) {
        this.source = source;
        this.destination = destination;
        this.pieceMoved = pieceMoved;
        this.capturedPiece = capturedPiece;
    }
}

class Board {
    static SIZE = 8;

    constructor() {
        this.cells = Array.from({ length: Board.SIZE }, () => new Array(Board.SIZE).fill(null));
        this.initialize();
    }

    initialize() {
        const backRank = [
            PieceType.ROOK, PieceType.KNIGHT, PieceType.BISHOP, PieceType.QUEEN,
            PieceType.KING, PieceType.BISHOP, PieceType.KNIGHT, PieceType.ROOK,
        ];
        for (let col = 0; col < Board.SIZE; col++) {
            this.cells[0][col] = Board._makePiece(backRank[col], Color.WHITE);
            this.cells[7][col] = Board._makePiece(backRank[col], Color.BLACK);
        }
        for (let col = 0; col < Board.SIZE; col++) {
            this.cells[1][col] = new Pawn(Color.WHITE);
            this.cells[6][col] = new Pawn(Color.BLACK);
        }
    }

    static _makePiece(type, color) {
        switch (type) {
            case PieceType.ROOK: return new Rook(color);
            case PieceType.KNIGHT: return new Knight(color);
            case PieceType.BISHOP: return new Bishop(color);
            case PieceType.QUEEN: return new Queen(color);
            case PieceType.KING: return new King(color);
            default: throw new Error('Unsupported back-rank piece type');
        }
    }

    isValidCell(row, col) {
        return row >= 0 && row < Board.SIZE && col >= 0 && col < Board.SIZE;
    }

    getPiece(row, col) {
        return this.cells[row][col];
    }

    setPiece(position, piece) {
        this.cells[position.row][position.col] = piece;
    }

    movePiece(source, destination) {
        const captured = this.getPiece(destination.row, destination.col);
        this.setPiece(destination, this.getPiece(source.row, source.col));
        this.setPiece(source, null);
        return captured;
    }

    findKing(color) {
        for (let row = 0; row < Board.SIZE; row++) {
            for (let col = 0; col < Board.SIZE; col++) {
                const piece = this.cells[row][col];
                if (piece && piece.type === PieceType.KING && piece.color === color) {
                    return { row, col };
                }
            }
        }
        return null;
    }

    allPieces(color) {
        const result = [];
        for (let row = 0; row < Board.SIZE; row++) {
            for (let col = 0; col < Board.SIZE; col++) {
                const piece = this.cells[row][col];
                if (piece && piece.color === color) {
                    result.push({ position: { row, col }, piece });
                }
            }
        }
        return result;
    }

    printBoard() {
        for (let row = Board.SIZE - 1; row >= 0; row--) {
            let line = `${row + 1} `;
            for (let col = 0; col < Board.SIZE; col++) {
                const piece = this.cells[row][col];
                line += piece ? piece.symbol() : '.';
                if (col !== Board.SIZE - 1) line += ' ';
            }
            console.log(line);
        }
        console.log('  a b c d e f g h');
    }
}

class Player {
    constructor(name, color) {
        this.name = name;
        this.color = color;
    }
}

function parseSquare(square) {
    const col = square.charCodeAt(0) - 'a'.charCodeAt(0);
    const row = parseInt(square[1], 10) - 1;
    return { row, col };
}

class IllegalMoveError extends Error {}

class Game {
    constructor(whiteName = 'White', blackName = 'Black') {
        this.board = new Board();
        this.white = new Player(whiteName, Color.WHITE);
        this.black = new Player(blackName, Color.BLACK);
        this.currentPlayer = this.white;
        this.status = GameStatus.ONGOING;
        // Undo stack: last move played sits at the end, ready to be reversed.
        this.history = [];
    }

    switchTurn() {
        this.currentPlayer = this.currentPlayer === this.white ? this.black : this.white;
    }

    isSquareAttacked(position, byColor) {
        for (const { position: piecePosition, piece } of this.board.allPieces(byColor)) {
            const attacked = piece.getAttackedSquares(piecePosition, this.board);
            if (attacked.some((square) => samePosition(square, position))) {
                return true;
            }
        }
        return false;
    }

    isCheck(color) {
        const kingPosition = this.board.findKing(color);
        if (!kingPosition) return false;
        return this.isSquareAttacked(kingPosition, opposite(color));
    }

    _hasLegalMove(color) {
        for (const { position: source, piece } of this.board.allPieces(color)) {
            for (const destination of piece.getPossibleMoves(source, this.board)) {
                const captured = this.board.movePiece(source, destination);
                const stillInCheck = this.isCheck(color);
                this.board.setPiece(source, piece);
                this.board.setPiece(destination, captured);
                if (!stillInCheck) return true;
            }
        }
        return false;
    }

    isCheckmate(color) {
        return this.isCheck(color) && !this._hasLegalMove(color);
    }

    isStalemate(color) {
        return !this.isCheck(color) && !this._hasLegalMove(color);
    }

    move(source, destination) {
        if (this.status === GameStatus.CHECKMATE || this.status === GameStatus.STALEMATE) {
            throw new IllegalMoveError('Game is already over');
        }

        const piece = this.board.getPiece(source.row, source.col);
        if (!piece || piece.color !== this.currentPlayer.color) {
            throw new IllegalMoveError('No movable piece of the current player at source');
        }

        const possible = piece.getPossibleMoves(source, this.board);
        if (!possible.some((square) => samePosition(square, destination))) {
            throw new IllegalMoveError('Illegal move for this piece');
        }

        let captured = this.board.movePiece(source, destination);
        if (this.isCheck(this.currentPlayer.color)) {
            // Undo the tentative move: it would leave (or keep) our own king in check.
            this.board.setPiece(source, piece);
            this.board.setPiece(destination, captured);
            throw new IllegalMoveError('Move leaves own king in check');
        }

        // The Move record keeps `piece` (pre-promotion) as pieceMoved so
        // undo() can restore the pawn even after we replace the board
        // occupant with a queen below.
        const performedMove = new Move(source, destination, piece, captured);
        this.history.push(performedMove);

        if (piece.type === PieceType.PAWN && (destination.row === 0 || destination.row === Board.SIZE - 1)) {
            this.board.setPiece(destination, new Queen(piece.color));
        }

        this.switchTurn();
        this._recomputeStatus();
        return performedMove;
    }

    // Pops the most recent move and restores the board exactly as it was
    // before that move, including undoing promotions (pieceMoved is always
    // the pre-promotion piece) and restoring captures.
    undo() {
        if (this.history.length === 0) {
            throw new IllegalMoveError('No moves to undo');
        }
        const lastMove = this.history.pop();
        this.board.setPiece(lastMove.source, lastMove.pieceMoved);
        this.board.setPiece(lastMove.destination, lastMove.capturedPiece);
        this.switchTurn();
        this._recomputeStatus();
    }

    _recomputeStatus() {
        const toMove = this.currentPlayer.color;
        if (this.isCheckmate(toMove)) {
            this.status = GameStatus.CHECKMATE;
        } else if (this.isCheck(toMove)) {
            this.status = GameStatus.CHECK;
        } else if (this.isStalemate(toMove)) {
            this.status = GameStatus.STALEMATE;
        } else {
            this.status = GameStatus.ONGOING;
        }
    }
}

function main() {
    const game = new Game('Alice', 'Bob');

    // Fool's Mate -- fastest possible checkmate, four half-moves.
    const moves = ['f2f3', 'e7e5', 'g2g4', 'd8h4'];
    for (const moveStr of moves) {
        const source = parseSquare(moveStr.slice(0, 2));
        const destination = parseSquare(moveStr.slice(2, 4));
        const mover = game.currentPlayer.name;
        game.move(source, destination);
        console.log(`${mover} played ${moveStr} -> status: ${game.status}`);
    }

    console.log();
    game.board.printBoard();
    console.log(`\nStatus before undo: ${game.status}`);

    console.log('\n-- Undoing the checkmating move --');
    game.undo();
    console.log(`Status after undo: ${game.status} (back to move for ${game.currentPlayer.name})`);
    game.board.printBoard();
}

if (require.main === module) {
    main();
}

module.exports = {
    Color, PieceType, GameStatus, Piece, Rook, Bishop, Queen, Knight, King, Pawn,
    Move, Board, Player, Game, IllegalMoveError, parseSquare,
};

File: Python/Interview_problems/Chess_Game.py
"""
================================================================================
LLD: Chess Game
================================================================================

Functional Requirements:
    1. Support a standard 8x8 board.
    2. Enforce piece-specific movement and capture rules for every piece type
       (KING, QUEEN, ROOK, BISHOP, KNIGHT, PAWN).
    3. Detect check, checkmate, and stalemate after every move.
    4. Maintain move history to support undo.

Non-Functional Requirements:
    1. Move validation latency should be small: each call generates a
       piece's candidate moves in O(board size), never scanning the whole
       game tree.
    2. Extensibility: adding a new piece type or a variant rule (e.g.
       castling) should not require touching unrelated piece classes.

--------------------------------------------------------------------------
Design
--------------------------------------------------------------------------
Strategy pattern for movement:
    Every piece knows only its OWN movement rule (Piece.get_possible_moves).
    Board/Game never special-case "if this is a bishop do X" -- they just
    call piece.get_possible_moves(...) polymorphically. Adding a new piece
    type means adding one new subclass, nothing else changes.

Command-style Move objects for undo:
    Each executed Move records exactly what it takes to reverse itself: the
    piece that was AT the source square before the move (so promotion can
    be undone back to a pawn) and whatever piece occupied the destination
    square before it was captured (possibly None). undo() pops the last
    Move and replays those two facts backwards -- it never needs to
    recompute anything from scratch.

Core Entities:
    Game             -- orchestrates turns, applies/undoes moves, tracks status
    Board            -- 8x8 grid of Piece references + board-level queries
    Piece (abstract)  -- Color, PieceType, get_possible_moves() [King, Queen,
                         Rook, Bishop, Knight, Pawn]
    Move             -- source, destination, piece_moved (pre-move), captured_piece
    Player           -- name + color
    GameStatus       -- ONGOING, CHECK, CHECKMATE, STALEMATE

Out of scope: castling, en passant, draw by repetition/50-move rule.
================================================================================
"""

from abc import ABC, abstractmethod
from enum import Enum, auto
from typing import List, Optional, Tuple

Position = Tuple[int, int]  # (row, col); row 0 = rank 1 (White's back rank), col 0 = file 'a'.


class Color(Enum):
    WHITE = auto()
    BLACK = auto()

    def opposite(self) -> "Color":
        return Color.BLACK if self == Color.WHITE else Color.WHITE


class PieceType(Enum):
    KING = auto()
    QUEEN = auto()
    ROOK = auto()
    BISHOP = auto()
    KNIGHT = auto()
    PAWN = auto()


class GameStatus(Enum):
    ONGOING = auto()
    CHECK = auto()
    CHECKMATE = auto()
    STALEMATE = auto()


class Piece(ABC):
    def __init__(self, color: Color, piece_type: PieceType):
        self.color = color
        self.type = piece_type

    @abstractmethod
    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        """Pseudo-legal moves: obeys this piece's movement/blocking/capture
        rules, but ignores whether the move leaves the mover's own king in
        check -- Game.move() layers that safety check on top."""
        raise NotImplementedError

    def get_attacked_squares(self, position: Position, board: "Board") -> List[Position]:
        """Squares this piece threatens; used for check detection. Identical
        to get_possible_moves for every piece except the pawn, whose
        diagonal attacks threaten a square whether or not an enemy is
        standing on it yet (a plain move-list wouldn't include an empty
        diagonal square)."""
        return self.get_possible_moves(position, board)

    def symbol(self) -> str:
        letters = {
            PieceType.KING: "K", PieceType.QUEEN: "Q", PieceType.ROOK: "R",
            PieceType.BISHOP: "B", PieceType.KNIGHT: "N", PieceType.PAWN: "P",
        }
        letter = letters[self.type]
        return letter if self.color == Color.WHITE else letter.lower()

    def _slide(self, position: Position, board: "Board", directions: List[Position]) -> List[Position]:
        """Shared helper for the sliding pieces (rook/bishop/queen): walk
        each direction until the edge of the board, a friendly piece (stop
        before it), or an enemy piece (stop after capturing it)."""
        moves = []
        row, col = position
        for d_row, d_col in directions:
            r, c = row + d_row, col + d_col
            while board.is_valid_cell((r, c)):
                occupant = board.get_piece((r, c))
                if occupant is None:
                    moves.append((r, c))
                else:
                    if occupant.color != self.color:
                        moves.append((r, c))
                    break
                r, c = r + d_row, c + d_col
        return moves


class Rook(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.ROOK)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        return self._slide(position, board, [(1, 0), (-1, 0), (0, 1), (0, -1)])


class Bishop(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.BISHOP)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        return self._slide(position, board, [(1, 1), (1, -1), (-1, 1), (-1, -1)])


class Queen(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.QUEEN)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        return self._slide(position, board, [
            (1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1),
        ])


class Knight(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.KNIGHT)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        row, col = position
        moves = []
        for d_row, d_col in [(2, 1), (2, -1), (-2, 1), (-2, -1), (1, 2), (1, -2), (-1, 2), (-1, -2)]:
            candidate = (row + d_row, col + d_col)
            if board.is_valid_cell(candidate):
                occupant = board.get_piece(candidate)
                if occupant is None or occupant.color != self.color:
                    moves.append(candidate)
        return moves


class King(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.KING)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        row, col = position
        moves = []
        for d_row in (-1, 0, 1):
            for d_col in (-1, 0, 1):
                if d_row == 0 and d_col == 0:
                    continue
                candidate = (row + d_row, col + d_col)
                if board.is_valid_cell(candidate):
                    occupant = board.get_piece(candidate)
                    if occupant is None or occupant.color != self.color:
                        moves.append(candidate)
        return moves


class Pawn(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.PAWN)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        row, col = position
        direction = 1 if self.color == Color.WHITE else -1
        start_row = 1 if self.color == Color.WHITE else 6
        moves = []

        one_step = (row + direction, col)
        if board.is_valid_cell(one_step) and board.get_piece(one_step) is None:
            moves.append(one_step)
            two_step = (row + 2 * direction, col)
            if row == start_row and board.get_piece(two_step) is None:
                moves.append(two_step)

        for d_col in (-1, 1):
            capture = (row + direction, col + d_col)
            if board.is_valid_cell(capture):
                occupant = board.get_piece(capture)
                if occupant is not None and occupant.color != self.color:
                    moves.append(capture)
        return moves

    def get_attacked_squares(self, position: Position, board: "Board") -> List[Position]:
        row, col = position
        direction = 1 if self.color == Color.WHITE else -1
        squares = []
        for d_col in (-1, 1):
            candidate = (row + direction, col + d_col)
            if board.is_valid_cell(candidate):
                squares.append(candidate)
        return squares


class Move:
    """A self-contained "command": records the two facts needed to reverse
    itself (piece_moved -- as it was BEFORE this move, so a promoted queen
    can be undone back into a pawn -- and captured_piece, possibly None),
    so Game.undo() never has to recompute board state from scratch."""

    def __init__(self, source: Position, destination: Position,
                 piece_moved: Piece, captured_piece: Optional[Piece]):
        self.source = source
        self.destination = destination
        self.piece_moved = piece_moved
        self.captured_piece = captured_piece


class Board:
    SIZE = 8

    def __init__(self):
        self.cells: List[List[Optional[Piece]]] = [[None] * Board.SIZE for _ in range(Board.SIZE)]
        self.initialize()

    def initialize(self) -> None:
        back_rank = [PieceType.ROOK, PieceType.KNIGHT, PieceType.BISHOP, PieceType.QUEEN,
                     PieceType.KING, PieceType.BISHOP, PieceType.KNIGHT, PieceType.ROOK]
        piece_classes = {
            PieceType.ROOK: Rook, PieceType.KNIGHT: Knight, PieceType.BISHOP: Bishop,
            PieceType.QUEEN: Queen, PieceType.KING: King,
        }
        for col, piece_type in enumerate(back_rank):
            self.cells[0][col] = piece_classes[piece_type](Color.WHITE)
            self.cells[7][col] = piece_classes[piece_type](Color.BLACK)
        for col in range(Board.SIZE):
            self.cells[1][col] = Pawn(Color.WHITE)
            self.cells[6][col] = Pawn(Color.BLACK)

    def is_valid_cell(self, position: Position) -> bool:
        row, col = position
        return 0 <= row < Board.SIZE and 0 <= col < Board.SIZE

    def get_piece(self, position: Position) -> Optional[Piece]:
        row, col = position
        return self.cells[row][col]

    def set_piece(self, position: Position, piece: Optional[Piece]) -> None:
        row, col = position
        self.cells[row][col] = piece

    def move_piece(self, source: Position, destination: Position) -> Optional[Piece]:
        captured = self.get_piece(destination)
        self.set_piece(destination, self.get_piece(source))
        self.set_piece(source, None)
        return captured

    def find_king(self, color: Color) -> Optional[Position]:
        for row in range(Board.SIZE):
            for col in range(Board.SIZE):
                piece = self.cells[row][col]
                if piece is not None and piece.type == PieceType.KING and piece.color == color:
                    return (row, col)
        return None

    def all_pieces(self, color: Color) -> List[Tuple[Position, Piece]]:
        result = []
        for row in range(Board.SIZE):
            for col in range(Board.SIZE):
                piece = self.cells[row][col]
                if piece is not None and piece.color == color:
                    result.append(((row, col), piece))
        return result

    def print_board(self) -> None:
        for row in range(Board.SIZE - 1, -1, -1):
            rank = [self.cells[row][col].symbol() if self.cells[row][col] else "." for col in range(Board.SIZE)]
            print(row + 1, " ".join(rank))
        print("  " + " ".join("abcdefgh"))


class Player:
    def __init__(self, name: str, color: Color):
        self.name = name
        self.color = color


def parse_square(square: str) -> Position:
    col = ord(square[0]) - ord("a")
    row = int(square[1]) - 1
    return (row, col)


class IllegalMoveError(Exception):
    pass


class Game:
    def __init__(self, white_name: str = "White", black_name: str = "Black"):
        self.board = Board()
        self.white = Player(white_name, Color.WHITE)
        self.black = Player(black_name, Color.BLACK)
        self.current_player = self.white
        self.status = GameStatus.ONGOING
        # Undo stack: last move played sits at the end, ready to be reversed.
        self.history: List[Move] = []

    def switch_turn(self) -> None:
        self.current_player = self.black if self.current_player is self.white else self.white

    def is_square_attacked(self, position: Position, by_color: Color) -> bool:
        for piece_position, piece in self.board.all_pieces(by_color):
            if position in piece.get_attacked_squares(piece_position, self.board):
                return True
        return False

    def is_check(self, color: Color) -> bool:
        king_position = self.board.find_king(color)
        if king_position is None:
            return False
        return self.is_square_attacked(king_position, color.opposite())

    def _has_legal_move(self, color: Color) -> bool:
        for source, piece in self.board.all_pieces(color):
            for destination in piece.get_possible_moves(source, self.board):
                captured = self.board.move_piece(source, destination)
                still_in_check = self.is_check(color)
                self.board.set_piece(source, piece)
                self.board.set_piece(destination, captured)
                if not still_in_check:
                    return True
        return False

    def is_checkmate(self, color: Color) -> bool:
        return self.is_check(color) and not self._has_legal_move(color)

    def is_stalemate(self, color: Color) -> bool:
        return not self.is_check(color) and not self._has_legal_move(color)

    def move(self, source: Position, destination: Position) -> Move:
        if self.status in (GameStatus.CHECKMATE, GameStatus.STALEMATE):
            raise IllegalMoveError("Game is already over")

        piece = self.board.get_piece(source)
        if piece is None or piece.color != self.current_player.color:
            raise IllegalMoveError("No movable piece of the current player at source")

        if destination not in piece.get_possible_moves(source, self.board):
            raise IllegalMoveError("Illegal move for this piece")

        captured = self.board.move_piece(source, destination)
        if self.is_check(self.current_player.color):
            # Undo the tentative move: it would leave (or keep) our own king in check.
            self.board.set_piece(source, piece)
            self.board.set_piece(destination, captured)
            raise IllegalMoveError("Move leaves own king in check")

        # The Move record keeps `piece` (pre-promotion) as piece_moved so
        # undo() can restore the pawn even after we replace the board
        # occupant with a queen below.
        performed_move = Move(source, destination, piece, captured)
        self.history.append(performed_move)

        if piece.type == PieceType.PAWN and destination[0] in (0, Board.SIZE - 1):
            self.board.set_piece(destination, Queen(piece.color))

        self.switch_turn()
        self._recompute_status()
        return performed_move

    def undo(self) -> None:
        """Pops the most recent move and restores the board exactly as it
        was before that move, including undoing promotions (piece_moved is
        always the pre-promotion piece) and restoring captures."""
        if not self.history:
            raise IllegalMoveError("No moves to undo")
        last_move = self.history.pop()
        self.board.set_piece(last_move.source, last_move.piece_moved)
        self.board.set_piece(last_move.destination, last_move.captured_piece)
        self.switch_turn()
        self._recompute_status()

    def _recompute_status(self) -> None:
        to_move = self.current_player.color
        if self.is_checkmate(to_move):
            self.status = GameStatus.CHECKMATE
        elif self.is_check(to_move):
            self.status = GameStatus.CHECK
        elif self.is_stalemate(to_move):
            self.status = GameStatus.STALEMATE
        else:
            self.status = GameStatus.ONGOING


def main() -> None:
    game = Game("Alice", "Bob")

    # Fool's Mate -- fastest possible checkmate, four half-moves.
    moves = ["f2f3", "e7e5", "g2g4", "d8h4"]
    for move_str in moves:
        source = parse_square(move_str[0:2])
        destination = parse_square(move_str[2:4])
        mover = game.current_player.name
        game.move(source, destination)
        print(f"{mover} played {move_str} -> status: {game.status.name}")

    print()
    game.board.print_board()
    print(f"\nStatus before undo: {game.status.name}")

    print("\n-- Undoing the checkmating move --")
    game.undo()
    print(f"Status after undo: {game.status.name} (back to move for {game.current_player.name})")
    game.board.print_board()


if __name__ == "__main__":
    main()
